### PR TITLE
Custom http header upload

### PR DESCRIPTION
**Pull request by fireworkmarks**

### CODE REVIEW

File: api/api-router.go [hunk 1]
```diff
@@ -87,6 +87,8 @@ func RegisterAPIRouter(mux *router.Router, api ObjectAPIHandlers) {
 		bucket.Methods("GET").Path("/{object:.+}").HandlerFunc(api.GetObjectHandler)
 		// DeleteObject
 		bucket.Methods("DELETE").Path("/{object:.+}").HandlerFunc(api.DeleteObjectHandler)
+		// POSTObjectCustomAttributes
```
Comment (cxt90730) [inline api/api-router.go]: .

File: api/datatype/metadata.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	"encoding/xml"
+	. "github.com/journeymidnight/yig/error"
+	"github.com/dustin/go-humanize"
+	"github.com/journeymidnight/yig/helper"
+	"io"
+	"io/ioutil"
+)
+
+const (
+	MaxObjectMetaConfigurationSize = 20 * humanize.KiByte
```
Comment (cxt90730) [inline api/datatype/metadata.go]: 2kb

File: api/datatype/metadata.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	Headers   Headers  `xml:"Headers,omitempty"`
+}
+
+type Headers struct {
+	XMLName         xml.Name        `xml:"Headers"`
+	HeaderCommon    HeaderCommon    `xml:"HeaderCommon"`
+	HeaderCustomize HeaderCustomize `xml:"HeaderCustomize"`
+}
+
+type HeaderCommon struct {
```
Comment (cxt90730) [inline api/datatype/metadata.go]: remove one

File: api/datatype/metadata.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+	"response-cache-control":       "Cache-Control",
+	"response-content-disposition": "Content-Disposition",
+	"response-content-language":    "Content-Language",
+	"response-content-encoding":    "Content-Encoding",
+}
+
+type MetaConfiguration struct {
+	XMLName   xml.Name `xml:"MetaConfiguration"`
+	Xmlns     string   `xml:"Xmlns,attr,omitempty"`
+	VersionID string   `xml:"VersionID"`
```
Comment (cxt90730) [inline api/datatype/metadata.go]: empty

File: storage/object.go [hunk 5]
```diff
@@ -651,6 +651,33 @@ func (yig *YigStorage) PutObject(bucketName string, objectName string, credentia
 	return result, nil
 }
 
+func (yig *YigStorage) PutObjectMeta(targetObject *meta.Object, credential common.Credential) (err error) {
+	bucket, err := yig.MetaStorage.GetBucket(targetObject.BucketName, true)
```
Comment (cxt90730) [inline storage/object.go]: use context

File: api/object-handlers.go [hunk 6]
```diff
@@ -1025,6 +1015,49 @@ func (api ObjectAPIHandlers) AppendObjectHandler(w http.ResponseWriter, r *http.
...
 }
 
+func (api ObjectAPIHandlers) PutObjectMeta(w http.ResponseWriter, r *http.Request) {
+	ctx := getRequestContext(r)
+	logger := ctx.Logger
+	vars := mux.Vars(r)
+	bucketName := vars["bucket"]
+	objectName := vars["object"]
+
+	logger.Info("Put object attributes:", bucketName, objectName)
```
Comment (fireworkmarks) [inline api/object-handlers.go]: 。

File: error/api-errors.go [hunk 7]
```diff
@@ -116,6 +116,7 @@ const (
 	ErrInvalidPosition
 	ErrObjectNotAppendable
 	ErrPositionNotEqualToLength
+	ErrMetadataHead
```
Comment (cxt90730) [inline error/api-errors.go]: Header

File: api/datatype/metadata.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+		return metaData, err
+	}
+	size := len(metaBuffer)
+	if size > MaxObjectMetaConfigurationSize {
+		return metaData, ErrEntityTooLarge
+	}
+	err = xml.Unmarshal(metaBuffer, metaConfig)
+	if err != nil {
+		helper.Logger.Error("Unable to parse metadata setting XML body:", err)
+		return metaData, ErrMalformedWebsiteConfiguration
```
Comment (cxt90730) [inline api/datatype/metadata.go]: .

File: api/datatype/metadata.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+	VersionId string
+	Data      map[string]string
+}
+
+func (w *MetaConfiguration) Validate() (metaData MetaData, error error) {
+	if w == nil {
+		return metaData, ErrEmptyEntity
+	}
+
+	if w.Headers.MetaData != nil {
```
Comment (cxt90730) [inline api/datatype/metadata.go]: len != 0

File: api/datatype/metadata.go [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+		for _, reqHead := range w.Headers.MetaData {
+			validMeta := strings.HasPrefix(reqHead.Key, CustomizeMetadataHead)
```
Comment (cxt90730) [inline api/datatype/metadata.go]: customizeMeta := strings.HasPrefix(reqHead.Key, CustomizeMetadataHeader)

File: api/datatype/metadata.go [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+	"strings"
+)
+
+const (
+	MaxObjectMetaConfigurationSize = 2 * humanize.KiByte
+	CustomizeMetadataHead          = "X-Amz-Meta-"
+)
+
+// supportedGetReqParams - supported request parameters for GET presigned request.
+var SupportedGetReqParams = map[string]string{
```
Comment (cxt90730) [inline api/datatype/metadata.go]: return

File: api/datatype/metadata.go [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+type MetaData struct {
+	VersionId string
+	Data      map[string]string
+}
+
+func (w *MetaConfiguration) Validate() (metaData MetaData, error error) {
+	if w == nil {
+		return metaData, ErrEmptyEntity
+	}
+
```
Comment (cxt90730) [inline api/datatype/metadata.go]: if w.Headers == nil

File: api/datatype/metadata.go [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+	Key     string   `xml:"Key"`
+	Value   string   `xml:"Value"`
+}
+
+type MetaData struct {
+	VersionId string
+	Data      map[string]string
+}
+
+func (w *MetaConfiguration) Validate() (metaData MetaData, error error) {
```
Comment (cxt90730) [inline api/datatype/metadata.go]: parse()

File: api/datatype/metadata.go [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+	Data      map[string]string
+}
+
+func (w *MetaConfiguration) parse() (MetaDataReq, error) {
+	metaDataReq := MetaDataReq{}
+	metaDataReq.Data = map[string]string{}
+	if w == nil {
+		return metaDataReq, ErrEmptyEntity
+	}
+
```
Comment (cxt90730) [inline api/datatype/metadata.go]: Conf require
Headers not require
Meta require if Headers exist

File: api/datatype/metadata.go [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+	if w.Headers == nil {
+		return metaDataReq, ErrEmptyEntity
+	}
+
+	if len(w.Headers.MetaData) != 0 {
+		for _, reqHeader := range w.Headers.MetaData {
+			validMeta := strings.HasPrefix(reqHeader.Key, CustomizeMetadataHead)
```
Comment (cxt90730) [inline api/datatype/metadata.go]: .

**Pull request merged**